sys/net/gnrc/routing/rpl/srh/gnrc_rpl_srh.c

```diff
@@ -12,11 +12,84 @@
  * @file
  */
 
+#include <string.h>
+#include "net/gnrc/ipv6/netif.h"
 #include "net/gnrc/rpl/srh.h"
 
-ipv6_addr_t *gnrc_rpl_srh_next_hop(gnrc_rpl_srh_t *rh)
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+#define GNRC_RPL_SRH_PADDING(X)     ((X & 0xF0) >> 4)
+#define GNRC_RPL_SRH_COMPRE(X)      (X & 0x0F)
+#define GNRC_RPL_SRH_COMPRI(X)      ((X & 0xF0) >> 4)
+
+ipv6_addr_t *gnrc_rpl_srh_next_hop(ipv6_hdr_t *ipv6, gnrc_rpl_srh_t *rh)
 {
-    /* TODO */
+    if (rh->seg_left) {
+        uint8_t n = (((rh->len * 8) - GNRC_RPL_SRH_PADDING(rh->pad_resv) -
+                     (16 - GNRC_RPL_SRH_COMPRE(rh->compr))) /
+                     (16 - GNRC_RPL_SRH_COMPRI(rh->compr))) + 1;
+        ipv6_addr_t addr = ipv6->dst;
+        ipv6_addr_t tmp = ipv6->dst;
+        uint8_t i = 0;
+        uint8_t pref_elided = GNRC_RPL_SRH_COMPRI(rh->compr);
+        uint8_t addr_len = sizeof(ipv6_addr_t) - pref_elided;
+        uint8_t *addr_vec = (uint8_t *) (rh + 1);
+
+        DEBUG("RPL SRH: %" PRIu8 " addresses in the routing header\n", n);
+
+        if (rh->seg_left > n) {
+            DEBUG("RPL SRH: number of segments left > number of addresses - discard\n");
+            /* TODO ICMP Parameter Problem - Code 0 */
+            return NULL;
+        }
+
+        rh->seg_left--;
+        i = n - rh->seg_left;
+        memcpy(&addr.u8[pref_elided], &addr_vec[(i - 1) * addr_len], addr_len);
```

Comment:
We should check `addr_vec[(i - 1) * addr_len]` does not access outside of the IPv6 packet.